### DISCUSSION TITLE
Help with the new getAudioBuffer function that returns a promise

### DISCUSSION THREAD

**Discussion by Franck-El** [jspsych/jsPsych#1729] [Q&A] [answered]
Hello,

I have a question about a jsPsych experiment that I haven't send on pavlovia yet. In the newest version the getAudioBuffer function returns a promise, not the buffer as it did before. I used this function to make my own sound plugin because I do multi-talker situations so I need to load all sounds before calcutate their rms value and apply them some gain (in fact a SNR because one sound is +6 dB higher than the sum of the others). I think I will not be able to do that with the newest version of this function so I keep the ancient for the moment but wonder what can I do if I have to update in the future ? I know promises are prefered because the code doesn't have to wait for the result but in my case, as I do some sound processing it's a necessity to load the sound before... I think... I'm not sure because I'm not familiar with asynchronous code and promises so I may not understand how powerful it is and how to make full use of it.

**Comment by jodeleeuw** [comment, marked as answer]
Hi @Franck-El,

We use a `Promise` in order to guarantee that the audio buffer is loaded before the trial starts. The old version of the audio plugins required that users preload their audio files, which led to some confusion when folks didn't preload and were unsure why the audio wasn't playing. The `Promise` allows us to write the plugin in a way that we check that the audio has finished loading before trying to make use of it, and if it hasn't finished loading we can try to load it and wait.

If you look at the [audio-button-response](https://github.com/jspsych/jsPsych/blob/master/plugins/jspsych-audio-button-response.js#L86-L266) plugin as an example, you'll see that we put all of the code that needs to run after the audio is loaded into a function called `setupTrial`. This function is called when the `Promise` from `getAudioBuffer` resolves. 

To convert old plugin code, you can use a similar strategy. Just wrap everything your old plugin did inside a function, and then invoke that function in the `.then()` call from the Promise.

**Reply by Franck-El** [reply]
Thank you for this really quick answer ! I just looked at the [audio-button-response](https://github.com/jspsych/jsPsych/blob/master/plugins/jspsych-audio-button-response.js#L86-L266) plugin, it seems clear now. I have an other similar question. I do a lot a sound processing before playing it so I wonder if I shouldn't better to do all of this before the experiment rather than doing it between trials when I go through the plugins ? I ask because I have a fast connection and a good computer so there's no timing problem but I'm afraid that it would be the case for a lower configuration. Do you think I'm right to worry ? It's the first time I do an online experiment and I use JS so I don't know how powerful is a browser. All my debug sessions are under chrome but I used Firefox once when I wasn't on my computer and I my website crashed 2 times over the 12 or 13 times I tested. The error was "TypeError: "attempting to access detached ArrayBuffer" when I passed one of my sound to a function that return its rms value. I read that it comes from a problem of memory, that's why I worry.
